src/telegram/error.py

```diff
@@ -208,21 +213,43 @@ class RetryAfter(TelegramError):
        :attr:`retry_after` is now an integer to comply with the Bot API.
 
     Args:
-        retry_after (:obj:`int`): Time in seconds, after which the bot can retry the request.
+        retry_after (:obj:`int` | :class:`datetime.timedelta`): Time in seconds, after which the
+            bot can retry the request.
+
+            .. versionchanged:: NEXT.VERSION
+                |time-period-input|
 
     Attributes:
-        retry_after (:obj:`int`): Time in seconds, after which the bot can retry the request.
+        retry_after (:obj:`int` | :class:`datetime.timedelta`): Time in seconds, after which the
+            bot can retry the request.
+
+            .. deprecated:: NEXT.VERSION
+                |time-period-int-deprecated|
 
     """
 
-    __slots__ = ("retry_after",)
+    __slots__ = ("_retry_after",)
+
+    def __init__(self, retry_after: TimePeriod):
+        self._retry_after: dtm.timedelta = parse_period_arg(  # type: ignore[assignment]
```

Comment:
the type checker is right in this case

Reply:
you mean b/c the return value is `timedelta | None` instead of `timedelta`? If you want to improve type hinting for `parse_period_arg` for that, you can use `typing.overload` to say `None → None` and `int | float | td → td`

Reply:
yeah that would fix it.